packages/core/src/render3/di.ts

```diff
@@ -344,6 +344,110 @@ export function injectAttributeImpl(tNode: TNode, attrNameToInject: string): str
   return null;
 }
 
+/**
+ * Returns a `TNode` of the location where the current `LView` is inserted into.
+ *
+ * @param lView an `LView` that we want to find parent `TNode` for.
+ */
+function getParentTNode(lView: LView): TNode|null {
+  const tView = lView[TVIEW];
+  if (tView.type === TViewType.Embedded) {
+    ngDevMode && assertDefined(tView.declTNode, 'Embedded TNodes should have declaration parents.');
+    return tView.declTNode;
+  } else if (tView.type === TViewType.Component) {
+    // Components don't have `TView.declTNode` because each instance of component could be
+    // inserted in different location, hence `TView.declTNode` is meaningless.
+    // Falling back to `T_HOST` in case we cross component boundary.
+    return lView[T_HOST];
+  }
+  // Remaining TNode type is `TViewType.Root` which doesn't have a parent TNode.
+  return null;
+}
+
+/**
+ * Returns the parent TNode and matching LView for a given TNode.
+ *
+ * Either returns the immediate node host or recursively navigates up the tree
+ * until it finds the next valid parent node, if it exists, and returns that.
+ *
+ * @param tNode a `TNode` that we want to find the parent of.
+ * @param lView an `LView` that corresponds to the tNode param.
+ * @param flags an `InjectFlags` that we check to skip @Host.
+ * @returns a tuple of the parent tNode and the corresponding lView of that node or null
+ */
+function getParentTNodeAndLView(tNode: TNode, lView: LView, flags: InjectFlags): [TNode, LView]|
+    null {
+  ngDevMode && assertTNodeForTView(tNode, lView[TVIEW]);
+  let parentTNode = tNode.parent as TNode | null;
+  let parentLView = lView;
+
+  // If `tNode.parent` is `null`, it means we are crossing a component boundary.
+  // We should proceed only when there is no `Host` flag provided.
+  if (parentTNode === null && !(flags & InjectFlags.Host)) {
+    parentTNode = getParentTNode(lView);
+    if (!parentTNode) return null;
+
+    // In this case, a parent exists and is definitely an element. So it will definitely
+    // an existing lView as the declaration view, which is why we can assume it's defined.
+    ngDevMode && assertDefined(parentLView, 'Declaration LView should be defined');
```

Comment:
It looks like this assert should either follow the next line (after `parentLView` has been reassigned) or this should be 
```suggestion
    ngDevMode && assertDefined(lView[DECLARATION_VIEW], 'Declaration LView should be defined');
```

Reply:
Actually it looks like we just need to update the comment to be clearer. The `parentLView` needs to exist, which is assigned on line 382. Otherwise `lView[DECLARATION_VIEW]` would throw. Maybe it would be clearer to update the following line to use `parentLView` instead of `lView`.

Reply:
I think I see what you mean. My observation was based on the fact that `lView` is guaranteed to be defined, as the parameter type is non-nullable. And at this point the `parentLView` is known to be `lView`, which makes `parentLView` known to be defined as well. That is probably what the comment is trying to get across. It seemed strange to have an assert at this point in the logic, also considering the immediate reassignment of `parentLView` without needing its value (why would there be an assertion when it's value is not even used at that point?).

Reply:
Adding the assertion was mainly to cover our bases. You're right that it should be defined since lView is non-nullable.